esp32c3-hal/examples/hmac.rs

```diff
@@ -0,0 +1,106 @@
+//! Demonstrates the use of the HMAC peripheral and compares the speed of
+//! hardware-accelerated and pure software hashing.
+
+#![no_std]
+#![no_main]
+
+use esp32c3_hal::{
+    clock::ClockControl,
+    hmac::{Hmac, HmacPurpose, KeyId},
+    peripherals::Peripherals,
+    prelude::*,
+    system::Peripheral as PeripheralEnable,
+    systimer::SystemTimer,
+    timer::TimerGroup,
+    Rng, Rtc,
+};
+use esp_backtrace as _;
+use esp_println::{print, println};
+use hmac::{Hmac as HmacSw, Mac};
+use nb::block;
+use sha2::Sha256;
+
+type HmacSha256 = HmacSw<Sha256>;
+
+#[entry]
+fn main() -> ! {
+    let peripherals = Peripherals::take();
+    let mut system = peripherals.SYSTEM.split();
+    let clocks = ClockControl::boot_defaults(system.clock_control).freeze();
+
+    // Disable the RTC and TIMG watchdog timers
+    let mut rtc = Rtc::new(peripherals.RTC_CNTL);
+    let timer_group0 = TimerGroup::new(
+        peripherals.TIMG0,
+        &clocks,
+        &mut system.peripheral_clock_control,
+    );
+    let mut wdt0 = timer_group0.wdt;
+    let timer_group1 = TimerGroup::new(
+        peripherals.TIMG1,
+        &clocks,
+        &mut system.peripheral_clock_control,
+    );
+    let mut wdt1 = timer_group1.wdt;
+
+    let mut rng = Rng::new(peripherals.RNG);
+
+    // Disable MWDT and RWDT (Watchdog) flash boot protection
+    rtc.rwdt.disable();
+    wdt0.disable();
+    wdt1.disable();
+
+    // Set peripheral clock and clear the peripheral reset
+    system
+        .peripheral_clock_control
+        .enable(PeripheralEnable::Sha);
+    system
```

Comment:
This should be done in the driver constructor, I know your PR is for HMAC but if you could fix it for SHA at the same time that would be appreciated!

Reply:
Just for my knowledge, is there any specific reason that enabling should be done in the driver constructor ?
It make sense that the clock gating and reset logic should be handled "elsewhere" because it can lead to over consuming in embedded software or am I wrong ?

Reply:
I think it's a bit of a convention in this HAL and we weren't very consistent about this in the past.
For me the idea is to have the driver in a working state after constructing it. There would be other ways to ensure the user enabled the peripheral clock but those would be more complex.
Also, probably in future we should make sure to disable the peripheral clock when the driver gets dropped 🤔 

btw I really like the detailed documentation in this PR 👍

Reply:
Updated for HMAC and SHA. I also updated examples.